### PR TITLE
Addition of Random frequency mask transform

### PR DESCRIPTION
**Pull request by payo101**
**Modified file(s)**
- `lightly/transforms/__init__.py`: Added new transform imports
- `lightly/transforms/random_crop_and_flip_with_grid.py`: Resolved `mypy` issue for `Any` subclassing

**Created file(s)**
- `lightly/transforms/random_frequency_mask_transform.py`
- `tests/transforms/test_random_frequency_mask_transform.py`

**Closes**: https://github.com/lightly-ai/lightly/issues/1660

### CODE REVIEW

File: lightly/transforms/random_frequency_mask_transform.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from typing import Tuple
+
+import numpy as np
+import torch
+from torch import Tensor
+
+
+class RFMTransform:
```
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:8]: Let's use the full name as RFM is not a well known abbreviation:

```suggestion
class RandomFrequencyMaskTransform:
```

File: lightly/transforms/random_frequency_mask_transform.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    """2D Random Frequency Mask Transformation.
+
+    This transformation applies a binary mask on the fourier transform,
+    across all channels. k% of frequencies are set to 0 with this.
+    k ranges [0.01, 0.1)
```
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:13]: Let's make `k` configurable:

```
def __init__(self, k: Tuple[float, float] = (0.01, 0.2)) -> None:
    self.k = k
```

File: lightly/transforms/random_frequency_mask_transform.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    Input
+        - Tensor: RFFT of a 2D Image (C, H, W) C-> No. of Channels
+    Output
+        - Tensor: The masked RFFT of the image
+
+    """
+
+    def __call__(self, fft_image: Tensor) -> Tensor:
+        k = np.random.uniform(0.01, 0.1)
```
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:23]:
```suggestion
        k = np.random.uniform(low=self.k[0], high=self.k[1])
```

File: lightly/transforms/random_frequency_mask_transform.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        # Mask: (C, H, W)
+        mask = torch.ones_like(fft_image)
```
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:25]: I believe the random mask creation can be simplified a bit:

```suggestion
        # Create mask with k entries set to zero.
        mask = torch.rand(fft_image.shape, device=fft_image.device) > k

        # Do not mask zero frequency mode to retain majority of semantic information.
        # See Section 4 in https://arxiv.org/abs/2312.02205 for details.
        mask[:, 0, 0] = 1

        masked_image = fft_image * mask
        return masked_image
```

File: lightly/transforms/random_frequency_mask_transform.py (hunk 2 above)
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:13]:
```suggestion
    This transformation applies a binary mask on the fourier transform,
    across all channels. A proportion of k frequencies are set to 0 with this.
```

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! I left some comments, let me know what you think. The most important is to make `k` configurable.

File: lightly/transforms/random_frequency_mask_transform.py (hunk 4 above)
Comment (payo101) [reply lightly/transforms/random_frequency_mask_transform.py:25]: What I understood from the paper is that the same binary mask needs to be applied on every channel of the RFFT of the Image. Based on that I think we can implement this. Please do tell if my understanding of the transform is wrong. Then your suggested change works best. 

```
k = np.random.uniform(low=self.k[0], high=self.k[1])

# Every mask for every channel will have same frequencies being turned off i.e. being set to zero
mask_type = torch.rand(fft_image.shape[1:], device=fft_image.device) > k # mask_type: (H, W)
mask = [] # Mask: (C, H, W)


for c in range(fft_image.size(dim=0)):
    mask.append(mask_type)
    
mask = torch.stack(mask)
# To retain majority of the semantic information. Please refer https://arxiv.org/abs/2312.02205
mask[:, 0, 0] = 1
masked_frequency_spectrum_image = fft_image * mask

return masked_frequency_spectrum_image
```

Comment (guarin) [reply lightly/transforms/random_frequency_mask_transform.py:25]: Ah yes you are right!

In that case we can do the masking like this:
```
        # Create mask with k entries set to zero.
        mask = torch.rand(fft_image.shape[1:], device=fft_image.device) > k
        # Do not mask zero frequency mode to retain majority of semantic information.
        # See Section 4 in https://arxiv.org/abs/2312.02205 for details.
        mask[0, 0] = 1
        mask = mask.unsqueeze(0) # Add channel dimension.
        masked_image = fft_image * mask
        return masked_image
```

**Comment by payo101** [comment]
@guarin @philippmwirth I have made the changes, could you guys review the PR now?

File: lightly/transforms/random_frequency_mask_transform.py [hunk 5]
```diff
@@ -19,20 +18,23 @@ class RFMTransform:
 
     """
 
-    def __call__(self, fft_image: Tensor) -> Tensor:
-        k = np.random.uniform(0.01, 0.1)
-        # Mask: (C, H, W)
-        mask = torch.ones_like(fft_image)
+    def __init__(self, k: Tuple[float, float] = (0.01, 0.2)) -> None:
```
Comment (guarin) [inline lightly/transforms/random_frequency_mask_transform.py:21]:
```suggestion
    def __init__(self, k: Tuple[float, float] = (0.01, 0.1)) -> None:
```

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks for the update! There is only one thing left, the default for k should be `(0.01, 0.1)` as in the paper.

**Comment by payo101** [comment]
Done @guarin

**Review by guarin — APPROVED** [review]
LGTM! 🔥
